File: android/app/src/main/java/com/hcjeong/forestix/sensors/DistanceSmoother.kt
// Robust moving average for AR-estimated distances (Hampel-style) —
// 1:1 port of iOS Sensors/DistanceSmoother.swift; keep the maths identical.
//
// ARCore's plane hit-test distance flickers frame-to-frame — single-frame
// spikes plus low-frequency wander — where the depth-API path is stable.
// This smoother keeps a short sliding window of recent samples and
// publishes the mean of the samples that agree with the window median
// (outliers beyond max(floor, k·MAD) are excluded from the average).
// Spikes DO stay in the buffer: if a "spike" persists (the cruiser
// actually walked closer), the median follows it within a few frames, so
// genuine movement tracks while flicker never shows.
//
// Used by the AR (non-depth) distance paths only: the caliper distance
// and the live distance readout when the Depth API is unavailable.

package com.hcjeong.forestix.sensors

import android.os.SystemClock
import kotlin.math.abs

class DistanceSmoother(
    private val capacity: Int = 12,
    private val maxAgeMs: Long = 1200,
    private val madMultiplier: Double = 3.0,
    /// MAD of a genuinely stable signal can collapse to ~0, which would
    /// reject everything — this floor keeps a few cm of slack.
    private val toleranceFloorM: Double = 0.03,
) {

    private data class Sample(val value: Double, val timeMs: Long)

    private val buffer = ArrayDeque<Sample>()

    fun add(value: Double, timeMs: Long = SystemClock.elapsedRealtime()) {
        if (!value.isFinite() || value <= 0) return
        buffer.addLast(Sample(value, timeMs))
        prune(timeMs)
    }

    fun reset() = buffer.clear()

    /// Robust mean of the recent window; null when no recent samples.
    fun value(nowMs: Long = SystemClock.elapsedRealtime()): Double? {
        prune(nowMs)
        val last = buffer.lastOrNull() ?: return null
        val values = buffer.map { it.value }
        if (values.size < 3) return last.value

        val sorted = values.sorted()
        val median = sorted[sorted.size / 2]
        val deviations = values.map { abs(it - median) }.sorted()
        val mad = deviations[deviations.size / 2]
        val tolerance = maxOf(toleranceFloorM, madMultiplier * mad)
        val kept = values.filter { abs(it - median) <= tolerance }
        if (kept.isEmpty()) return median
        return kept.average()
    }

    private fun prune(nowMs: Long) {
        while (buffer.isNotEmpty() && nowMs - buffer.first().timeMs > maxAgeMs) {
            buffer.removeFirst()
        }
        while (buffer.size > capacity) buffer.removeFirst()
    }
}

File: TimberCruisingApp/Sensors/DistanceSmoother.swift
// Robust moving average for AR-estimated distances (Hampel-style).
//
// ARKit's estimated-plane raycast distance flickers frame-to-frame —
// single-frame spikes plus low-frequency wander — where the LiDAR path
// is stable. This smoother keeps a short sliding window of recent
// samples and publishes the mean of the samples that agree with the
// window median (outliers beyond max(floor, k·MAD) are excluded from
// the average). Spikes DO stay in the buffer: if a "spike" persists
// (the cruiser actually walked closer), the median follows it within a
// few frames, so genuine movement tracks while flicker never shows.
//
// Used by the AR (non-LiDAR) distance paths only: the caliper distance
// and the live distance readout in AR mode. Mirror of the Android
// sensors/DistanceSmoother.kt — keep the maths identical.

import Foundation

public final class DistanceSmoother {

    private struct Sample {
        let value: Double
        let time: TimeInterval
    }

    private var buffer: [Sample] = []
    private let capacity: Int
    private let maxAgeSec: TimeInterval
    private let madMultiplier: Double
    /// MAD of a genuinely stable signal can collapse to ~0, which would
    /// reject everything — this floor keeps a few cm of slack.
    private let toleranceFloorM: Double

    public init(capacity: Int = 12,
                maxAgeSec: TimeInterval = 1.2,
                madMultiplier: Double = 3.0,
                toleranceFloorM: Double = 0.03) {
        self.capacity = capacity
        self.maxAgeSec = maxAgeSec
        self.madMultiplier = madMultiplier
        self.toleranceFloorM = toleranceFloorM
    }

    public func add(_ value: Double,
                    at time: TimeInterval = ProcessInfo.processInfo.systemUptime) {
        guard value.isFinite, value > 0 else { return }
        buffer.append(Sample(value: value, time: time))
        prune(now: time)
    }

    public func reset() {
        buffer.removeAll(keepingCapacity: true)
    }

    /// Robust mean of the recent window; nil when no recent samples.
    public func value(now: TimeInterval = ProcessInfo.processInfo.systemUptime) -> Double? {
        prune(now: now)
        guard let last = buffer.last else { return nil }
        let values = buffer.map(\.value)
        guard values.count >= 3 else { return last.value }

        let sorted = values.sorted()
        let median = sorted[sorted.count / 2]
        let deviations = values.map { abs($0 - median) }.sorted()
        let mad = deviations[deviations.count / 2]
        let tolerance = max(toleranceFloorM, madMultiplier * mad)
        let kept = values.filter { abs($0 - median) <= tolerance }
        guard !kept.isEmpty else { return median }
        return kept.reduce(0, +) / Double(kept.count)
    }

    private func prune(now: TimeInterval) {
        buffer.removeAll { now - $0.time > maxAgeSec }
        if buffer.count > capacity {
            buffer.removeFirst(buffer.count - capacity)
        }
    }
}
